ThisAssembly.Git.targets fails to resolve branch name when working with worktrees

## Describe the Bug

`ThisAssembly.Git` fails during an MSBuild build when the repository is checked out as a Git worktree.

In a worktree, `.git` is a file containing a `gitdir:` pointer instead of a directory. However, `ThisAssembly.Git.targets` assumes that `$(RepositoryRoot).git` is always a directory and attempts to read `.git/HEAD` directly:

```xml
<_ThisAssemblyGitBranchName Condition="'$(_ThisAssemblyGitBranchName)' == ''">
  $([System.IO.File]::ReadAllText('$(RepositoryRoot).git/HEAD').Trim())
</_ThisAssemblyGitBranchName>
```

Consequently, branch resolution fails before the project can be compiled.

## Steps to Reproduce

1. Create a repository containing an SDK-style .NET project referencing `ThisAssembly.Git` 2.1.2.
2. Configure the repository to use .NET SDK 8.
3. Create a worktree:

```shell
git worktree add ../repository-worktree -b reproduce-thisassembly-worktree
```

4. Confirm that `.git` in the worktree is a file:

```shell
cat ../repository-worktree/.git
```

Example output:

```text
gitdir: ../original-repository/.git/worktrees/repository-worktree
```

5. Build the project from the worktree:

```shell
dotnet build path/to/Project.csproj
```

Alternatively, invoke the affected target directly:

```shell
dotnet msbuild path/to/Project.csproj \
  -t:InitializeGitInformation \
  -getProperty:RepositoryBranch
```

The build fails when `ThisAssembly.Git.targets` tries to read `.git/HEAD`.

## Expected Behavior

`ThisAssembly.Git` should resolve `RepositoryBranch` correctly when building from either:

- A regular Git checkout, where `.git` is a directory.
- A Git worktree, where `.git` is a file containing a `gitdir:` pointer.

For example, the package could use Git itself:

```shell
git rev-parse --abbrev-ref HEAD
```

Alternatively, it could parse the worktree's `.git` pointer and read `HEAD` from the referenced Git administrative directory.

At a minimum, the direct `.git/HEAD` fallback should verify that the file exists before calling `System.IO.File.ReadAllText`.

## Exception with Stack Trace

```text
ThisAssembly.Git.targets(79,7): error MSB4184:
The expression
"([System.IO.File]::ReadAllText(C:\path\to\repository-worktree\.git/HEAD))"
cannot be evaluated.

Could not find a part of the path
'C:\path\to\repository-worktree\.git\HEAD'.
```

The failing statement is:

```xml
<_ThisAssemblyGitBranchName Condition="'$(_ThisAssemblyGitBranchName)' == ''">
  $([System.IO.File]::ReadAllText('$(RepositoryRoot).git/HEAD').Trim())
</_ThisAssemblyGitBranchName>
```

## Version Info

- `ThisAssembly.Git`: 2.1.2
- .NET SDK: 8.0.422
- Target framework: .NET Framework 4.8 (`net48`)
- Operating system: Windows
- Git checkout type: linked worktree
- Build command: `dotnet msbuild` / `dotnet build`

## Additional Info

This appears to affect the fallback intended for .NET SDK 8. The target contains the following comment:

```xml
<!-- The above only works on .NET SDK 9.0+, so account for 8.0 implicit nuget package -->
```

The preceding SourceLink-based branch metadata was empty in this environment, causing the `.git/HEAD` fallback to execute.

As a temporary workaround, I added a repository-level target that populates `RepositoryBranch` before `InitializeGitInformation`:

```xml
<Project>
  <Target Name="ResolveThisAssemblyGitWorktreeBranch"
          BeforeTargets="InitializeGitInformation"
          Condition="'$(RepositoryBranch)' == '' and
                     Exists('$(RepositoryRoot).git') and
                     !Exists('$(RepositoryRoot).git/HEAD')">

    <Exec Command="git -C &quot;$(RepositoryRoot)&quot; rev-parse --abbrev-ref HEAD"
          ConsoleToMSBuild="true">
      <Output TaskParameter="ConsoleOutput"
              PropertyName="RepositoryBranch" />
    </Exec>
  </Target>
</Project>
```

With this workaround, `InitializeGitInformation` completes successfully and `RepositoryBranch` contains the worktree's current branch.

<!-- sl -->
[![Back this issue](https://sponsorlink.blob.core.windows.net/public/separator.png "Back this issue")](https://www.devlooped.com/SponsorLink/github/issues/?s=devlooped)
[![Back this issue](https://img.shields.io/badge/backed-%240-EA4AAA?logo=githubsponsors "Back this issue")](https://www.devlooped.com/SponsorLink/github/issues/?s=devlooped)
<!-- sl -->

--- a/src/ThisAssembly.Git/ThisAssembly.Git.targets
+++ b/src/ThisAssembly.Git/ThisAssembly.Git.targets
@@ -72,12 +72,21 @@
     <Warning Code="TA001" Text="A valid SourceLink provider does not seem to be installed for the current repository/project. Some values may be empty." Condition="'$(SourceLinkUrl)' == ''" />
     
     <PropertyGroup Condition="'$(RepositoryBranch)' == '' and '$(RepositoryRoot)' != ''">
-      <!-- If we didnd't populate from CI/ENV, we can still use the source control provider itself to get the branch name -->
+      <!-- If we didn't populate from CI/ENV, we can still use the source control provider itself to get the branch name -->
       <_ThisAssemblyGitBranchName>@(_ThisAssemblyGitSourceRoot -> '%(BranchName)')</_ThisAssemblyGitBranchName>
       <!-- The above only works on .NET SDK 9.0+, so account for 8.0 implicit nuget package -->
 
-      <_ThisAssemblyGitBranchName Condition="'$(_ThisAssemblyGitBranchName)' == ''">$([System.IO.File]::ReadAllText('$(RepositoryRoot).git/HEAD').Trim())</_ThisAssemblyGitBranchName>
-      <RepositoryBranch>$(_ThisAssemblyGitBranchName.TrimStart("ref:").Trim().Replace('refs/heads/', '').Replace('refs/tags/', ''))</RepositoryBranch>
+      <!-- Regular checkout: .git is a directory containing HEAD -->
+      <_ThisAssemblyGitHead Condition="'$(_ThisAssemblyGitBranchName)' == '' and Exists('$(RepositoryRoot).git/HEAD')">$(RepositoryRoot).git/HEAD</_ThisAssemblyGitHead>
+
+      <!-- Worktree / separate-git-dir / submodule: .git is a file with 'gitdir: &lt;path&gt;' -->
+      <_ThisAssemblyGitFile Condition="'$(_ThisAssemblyGitHead)' == '' and '$(_ThisAssemblyGitBranchName)' == '' and $([System.IO.File]::Exists('$(RepositoryRoot).git'))">$([System.IO.File]::ReadAllText('$(RepositoryRoot).git').Trim())</_ThisAssemblyGitFile>
+      <_ThisAssemblyGitDir Condition="'$(_ThisAssemblyGitFile)' != '' and $(_ThisAssemblyGitFile.StartsWith('gitdir:', StringComparison.OrdinalIgnoreCase))">$(_ThisAssemblyGitFile.Substring(7).Trim())</_ThisAssemblyGitDir>
+      <_ThisAssemblyGitDir Condition="'$(_ThisAssemblyGitDir)' != ''">$([MSBuild]::NormalizePath('$(RepositoryRoot)', '$(_ThisAssemblyGitDir)'))</_ThisAssemblyGitDir>
+      <_ThisAssemblyGitHead Condition="'$(_ThisAssemblyGitHead)' == '' and '$(_ThisAssemblyGitDir)' != '' and Exists('$(_ThisAssemblyGitDir)/HEAD')">$(_ThisAssemblyGitDir)/HEAD</_ThisAssemblyGitHead>
+
+      <_ThisAssemblyGitBranchName Condition="'$(_ThisAssemblyGitBranchName)' == '' and '$(_ThisAssemblyGitHead)' != ''">$([System.IO.File]::ReadAllText('$(_ThisAssemblyGitHead)').Trim())</_ThisAssemblyGitBranchName>
+      <RepositoryBranch Condition="'$(_ThisAssemblyGitBranchName)' != ''">$(_ThisAssemblyGitBranchName.TrimStart("ref:").Trim().Replace('refs/heads/', '').Replace('refs/tags/', ''))</RepositoryBranch>
     </PropertyGroup>
 
   </Target>

--- a/src/ThisAssembly.Tests/GitTests.cs
+++ b/src/ThisAssembly.Tests/GitTests.cs
@@ -1,0 +1,158 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace ThisAssemblyTests;
+
+/// <summary>
+/// Exercises the <c>InitializeGitInformation</c> fallback that reads HEAD when
+/// SourceLink does not provide <c>BranchName</c> (typical on .NET SDK 8).
+/// </summary>
+public record class GitTests(ITestOutputHelper Output)
+{
+    /// <summary />
+    [Fact]
+    public void ReadsBranchFromGitDirectory()
+    {
+        using var dir = new TempDirectory();
+        var repo = Path.Combine(dir.Path, "repo");
+        Directory.CreateDirectory(Path.Combine(repo, ".git"));
+        File.WriteAllText(Path.Combine(repo, ".git", "HEAD"), "ref: refs/heads/main\n");
+
+        Assert.Equal("main", EvaluateBranch(repo));
+    }
+
+    /// <summary />
+    [Fact]
+    public void ReadsBranchFromWorktreeGitDirFileWithRelativePath()
+    {
+        using var dir = new TempDirectory();
+        var main = Path.Combine(dir.Path, "main");
+        var worktree = Path.Combine(dir.Path, "worktree");
+        var gitDir = Path.Combine(main, ".git", "worktrees", "worktree");
+
+        Directory.CreateDirectory(gitDir);
+        Directory.CreateDirectory(worktree);
+        File.WriteAllText(Path.Combine(gitDir, "HEAD"), "ref: refs/heads/feature/worktree\n");
+
+        var relativeGitDir = Path.GetRelativePath(worktree, gitDir).Replace('\\', '/');
+        File.WriteAllText(Path.Combine(worktree, ".git"), $"gitdir: {relativeGitDir}\n");
+
+        Assert.Equal("feature/worktree", EvaluateBranch(worktree));
+    }
+
+    /// <summary />
+    [Fact]
+    public void ReadsBranchFromWorktreeGitDirFileWithAbsolutePath()
+    {
+        using var dir = new TempDirectory();
+        var main = Path.Combine(dir.Path, "main");
+        var worktree = Path.Combine(dir.Path, "worktree");
+        var gitDir = Path.Combine(main, ".git", "worktrees", "worktree");
+
+        Directory.CreateDirectory(gitDir);
+        Directory.CreateDirectory(worktree);
+        File.WriteAllText(Path.Combine(gitDir, "HEAD"), "ref: refs/heads/reproduce-thisassembly-worktree\n");
+
+        var absoluteGitDir = gitDir.Replace('\\', '/');
+        File.WriteAllText(Path.Combine(worktree, ".git"), $"gitdir: {absoluteGitDir}\n");
+
+        Assert.Equal("reproduce-thisassembly-worktree", EvaluateBranch(worktree));
+    }
+
+    /// <summary />
+    [Fact]
+    public void DoesNotFailWhenGitHeadIsMissing()
+    {
+        using var dir = new TempDirectory();
+        var repo = Path.Combine(dir.Path, "repo");
+        Directory.CreateDirectory(repo);
+
+        Assert.Equal("", EvaluateBranch(repo));
+    }
+
+    string EvaluateBranch(string repositoryRoot)
+    {
+        var targetsPath = Path.GetFullPath(Path.Combine(ThisAssembly.Git.Root, "src", "ThisAssembly.Git", "ThisAssembly.Git.targets"));
+        Assert.True(File.Exists(targetsPath), $"Could not find ThisAssembly.Git.targets at {targetsPath}");
+
+        var evalDir = Path.Combine(Path.GetTempPath(), "ThisAssemblyGitEval", Guid.NewGuid().ToString("N"));
+        Directory.CreateDirectory(evalDir);
+        try
+        {
+            var projectPath = Path.Combine(evalDir, "git.proj");
+            var outputPath = Path.Combine(evalDir, "branch.txt");
+            var root = repositoryRoot.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
+                + Path.DirectorySeparatorChar;
+
+            File.WriteAllText(projectPath, $$"""
+                <Project>
+                  <Target Name="InitializeSourceControlInformation" />
+                  <Import Project="{{Xml(targetsPath)}}" />
+                  <!-- Force the .git/HEAD fallback, as if CI/SourceLink did not populate the branch. -->
+                  <PropertyGroup>
+                    <RepositoryBranch></RepositoryBranch>
+                  </PropertyGroup>
+                  <ItemGroup>
+                    <SourceRoot Include="{{Xml(root)}}">
+                      <SourceControl>git</SourceControl>
+                    </SourceRoot>
+                  </ItemGroup>
+                  <Target Name="WriteBranch" DependsOnTargets="InitializeGitInformation">
+                    <WriteLinesToFile File="{{Xml(outputPath)}}" Lines="$(RepositoryBranch)" Overwrite="true" />
+                  </Target>
+                </Project>
+                """);
+
+            var psi = new ProcessStartInfo("dotnet")
+            {
+                Arguments = $"msbuild \"{projectPath}\" -nologo -v:m -t:WriteBranch",
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+                WorkingDirectory = evalDir,
+            };
+
+            using var process = Process.Start(psi) ?? throw new InvalidOperationException("Failed to start dotnet msbuild.");
+            var stdout = process.StandardOutput.ReadToEnd();
+            var stderr = process.StandardError.ReadToEnd();
+            process.WaitForExit();
+
+            Output.WriteLine(stdout);
+            if (!string.IsNullOrWhiteSpace(stderr))
+                Output.WriteLine(stderr);
+
+            if (process.ExitCode != 0)
+                throw new InvalidOperationException($"msbuild failed ({process.ExitCode}):{Environment.NewLine}{stdout}{stderr}");
+
+            return File.Exists(outputPath) ? File.ReadAllText(outputPath).Trim() : "";
+        }
+        finally
+        {
+            try { Directory.Delete(evalDir, recursive: true); } catch { }
+        }
+    }
+
+    static string Xml(string value) => value
+        .Replace("&", "&amp;")
+        .Replace("\"", "&quot;")
+        .Replace("'", "&apos;")
+        .Replace("<", "&lt;")
+        .Replace(">", "&gt;");
+
+    sealed class TempDirectory : IDisposable
+    {
+        public string Path { get; } = System.IO.Path.Combine(
+            System.IO.Path.GetTempPath(), "ThisAssemblyGitTests", Guid.NewGuid().ToString("N"));
+
+        public TempDirectory() => Directory.CreateDirectory(Path);
+
+        public void Dispose()
+        {
+            try { Directory.Delete(Path, recursive: true); } catch { }
+        }
+    }
+}

